pkg/cli/run.go

```diff
@@ -118,40 +137,89 @@ func (r *RunUI) run(ctx context.Context) {
 	go func() {
 		if err := service.Start(ctx, exec); err != nil {
 			r.err = err
-			r.done = true
 			return
 		}
 	}()
 
 	// XXX: We need to define a readiness check with each of our services,
 	// then wait here for the readiness check to pass.
 
-	r.id, err = runner.Initialize(ctx, r.fn, r.event, r.sm, q)
-	if err != nil {
-		r.err = err
-		return
-	}
-	if r.id == nil {
-		r.err = fmt.Errorf("no run id created")
-		return
-	}
+	var wg sync.WaitGroup
 
-	for !r.done {
-		var run state.State
+	for _, evt := range r.events {
+		wg.Add(1)
 
-		run, r.err = r.sm.Load(ctx, *r.id)
-		if r.err != nil {
-			return
-		}
+		go func(event event.Event) {
+			defer wg.Done()
 
-		meta := run.Metadata()
-		if meta.Pending == 0 {
-			r.duration = time.Since(meta.StartedAt)
-			r.done = true
-			return
-		}
-		<-time.After(time.Millisecond)
+			var runId *state.Identifier
+
+			runId, err = runner.Initialize(ctx, r.fn, event, r.sm, q)
+			if err != nil {
+				r.err = err
+				return
+			}
+			if runId == nil {
+				r.err = fmt.Errorf("no run id created")
+				return
+			}
+
+			done := false
+			seenOutput := []string{}
+
+			execution := RunUIExecution{
+				id:         runId,
+				done:       &done,
+				event:      &event,
+				seenOutput: &seenOutput,
+			}
+
+			r.runs = append(r.runs, execution)
+
+			for !*execution.done {
+				var run state.State
+
+				run, err = r.sm.Load(ctx, *runId)
+				if err != nil {
+					r.err = err
+					return
+				}
+
+				output := run.Actions()
+
+				for id := range output {
+					haveSeenKey := false
+
+					for _, seenKey := range seenOutput {
+						if seenKey == id {
+							haveSeenKey = true
+							break
+						}
+					}
+
+					if !haveSeenKey {
+						seenOutput = append(seenOutput, id)
+					}
+				}
+
+				hasErrors := len(run.Errors()) > 0
+
+				if run.Metadata().Pending == 0 || hasErrors {
+					if hasErrors {
+						r.err = fmt.Errorf("run failed")
+					}
+
+					*execution.done = true
+
+					return
+				}
+				<-time.After(time.Millisecond * 5)
```

Comment:
This feels clunky; I could use channels instead to ensure it's not needlessly looping.

Reply:
I understand what you mean and think that we could provide something that allows you to subscribe to function runs from the state store (eg. new state, and/or completion).

For now, we have to poll _somewhere_ to monitor state, so I'm okay with adding this here :)